FlexMeter.c

```diff
@@ -0,0 +1,203 @@
+#include "config.h"
+#include "FlexMeter.h"
+#include "Object.h"
+
+
+#define FLEX_CFG_FOLDER ".config/htop/FlexMeter"
+
+typedef struct {
+   char* name;
+   char* command;
+   char* type;
+   char* caption;
+   char* uiName;
+} _flex_meter;
+
+_flex_meter meter_list[METERS_LIST_SIZE];
+
+static int meters_count = 0;
+
+static const int DateMeter_attributes[] = {
+   FLEX
+};
+
+MeterClass* FlexMeter_class = NULL;
+
+static int check_for_meters(void);
+
+static bool parse_input(_flex_meter *meter, char* line)
+{
+   switch(line[0])
+   {
+      case 'n':
+         if (String_startsWith(line, "name=")) {
+            xAsprintf(&meter->uiName, "Flex: %s", line + 5);
+         }
+         break;
+      case 'c':
+         if (String_startsWith(line, "command=")) {
+            meter->command = xStrdup(line + 8);
+         } else if (String_startsWith(line, "caption=")) {
+            meter->caption = xStrdup(line + 8);
+         }
+         break;
+      case 't':
+         if (String_startsWith(line, "type=")) {
+            meter->type = xStrdup(line + 6);
+         }
+         break;
+      default:
+         return false;
+   }
+
+   return true;
+}
+
+static bool load_config(_flex_meter *meter, char* file)
+{
+   bool ret = false;
+   FILE* fp = fopen(file, "r");
+
+   if (fp != NULL) {
+      char* buff;
+      while ((buff = String_readLine(fp)) != NULL) {
+         ret = parse_input(meter, buff);
+         if (!ret) {
+            break;
+         }
+      }
+      free(buff);
+      buff = NULL;
+   }
+
+   fclose(fp);
+   return ret;
+}
+
+static int check_for_meters(void)
+{
+   char* path;
+   struct dirent* dir;
+   struct passwd* pw = getpwuid(getuid());
+   const char* homedir = pw->pw_dir;
+   char* home = NULL;
+   bool ret;
+
+   xAsprintf(&home, "%s/%s", homedir, FLEX_CFG_FOLDER);
+   struct stat fileStat;
+
+   if (stat(home, &fileStat) < 0) {
+      return -1;
+   }
+
+   uint32_t uid = getuid();
+
+   if ((fileStat.st_uid == uid) && S_ISDIR(fileStat.st_mode) && ((fileStat.st_mode & 0777) == 0700)) {
+      DIR* d = opendir(home);
+      if (d) {
+         while ((dir = readdir(d)) != NULL) {
+            if ( dir->d_name[0] == '.') {
+               /* We are ignoring all files starting with . like ".Template"
+                * and "." ".." directories
+                */
+               continue;
+            }
+
+            meter_list[meters_count].name = xStrdup(dir->d_name);
+            xAsprintf(&path, "%s/%s", home, dir->d_name);
+
+            if (access(path, R_OK | W_OK | X_OK) == 0) {
+               ret = load_config(&meter_list[meters_count], path);
+
+               if (ret && (meters_count < MAX_METERS_COUNT)) {
+                  meters_count++;
+               }
+            }
+            if (path != NULL) {
+               free(path);
+               path=NULL;
+            }
+         }
+         closedir(d);
+      }
+   }
+
+   if (home != NULL) {
+      free(home);
+      home = NULL;
+   }
```

Comment:
You can assume `free(NULL)` will do nothing. This behavior is required by ANSI C standard. (Operating systems that do not have this behavior are too old for htop to support.)

Reply:
To be precise: A libc where `free(NULL);` is anything but a no-op does not comply to the ISO C99 standard (section 7.20.3.2) and is thus outside the scope of what we support.

Reply:
Just to confirm which approach is preferred here with `if (ptr != NULL)` of without ?

Reply:
Without …

Reply:
My fault , initially I understood that this check is required.
I will remove them